From 7b750ab655eb268fa9364b4c0ad0bdca8778bf24 Mon Sep 17 00:00:00 2001
From: Sang Yan <sangyan@huawei.com>
Date: Tue, 27 Jul 2021 22:02:12 +0800
Subject: [PATCH] arm64: Reserve memory for quick kexec
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 7b750ab655eb268fa9364b4c0ad0bdca8778bf24
Modified-by-SEL: Yes, refreshed due to context change


hulk inclusion
category: feature
bugzilla: 48159
CVE: N/A

------------------------------

Reserve memory for quick kexec on arm64
with cmdline "quickkexec=".

Signed-off-by: Sang Yan <sangyan@huawei.com>
Reviewed-by: Chen Wandun <chenwandun@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Geliang Tang <geliang.tang@suse.com>
Signed-off-by: Kai Liu <kai.liu@suse.com>
---
 arch/arm64/kernel/setup.c |    6 ++++++
 arch/arm64/mm/init.c      |   42 ++++++++++++++++++++++++++++++++++++++++++
 2 files changed, 48 insertions(+)

--- a/arch/arm64/kernel/setup.c
+++ b/arch/arm64/kernel/setup.c
@@ -267,6 +267,12 @@ static void __init request_standard_reso
 		    crashk_res.end <= res->end)
 			request_resource(res, &crashk_res);
 #endif
+#ifdef CONFIG_QUICK_KEXEC
+		if (quick_kexec_res.end &&
+		    quick_kexec_res.start >= res->start &&
+		    quick_kexec_res.end <= res->end)
+			request_resource(res, &quick_kexec_res);
+#endif
 	}
 }
 
--- a/arch/arm64/mm/init.c
+++ b/arch/arm64/mm/init.c
@@ -154,6 +154,45 @@ static void __init reserve_elfcorehdr(vo
 }
 #endif /* CONFIG_CRASH_DUMP */
 
+#ifdef CONFIG_QUICK_KEXEC
+static int __init parse_quick_kexec(char *p)
+{
+	if (!p)
+		return 0;
+
+	quick_kexec_res.end = PAGE_ALIGN(memparse(p, NULL));
+
+	return 0;
+}
+early_param("quickkexec", parse_quick_kexec);
+
+static void __init reserve_quick_kexec(void)
+{
+	unsigned long long mem_start, mem_len;
+
+	mem_len = quick_kexec_res.end;
+	if (mem_len == 0)
+		return;
+
+	/* Current arm64 boot protocol requires 2MB alignment */
+	mem_start = memblock_find_in_range(0, arm64_dma_phys_limit,
+			mem_len, SZ_2M);
+	if (mem_start == 0) {
+		pr_warn("cannot allocate quick kexec mem (size:0x%llx)\n",
+			mem_len);
+		quick_kexec_res.end = 0;
+		return;
+	}
+
+	memblock_reserve(mem_start, mem_len);
+	pr_info("quick kexec mem reserved: 0x%016llx - 0x%016llx (%lld MB)\n",
+		mem_start, mem_start + mem_len,	mem_len >> 20);
+
+	quick_kexec_res.start = mem_start;
+	quick_kexec_res.end = mem_start + mem_len - 1;
+}
+#endif
+
 /*
  * Return the maximum physical address for a zone accessible by the given bits
  * limit. If DRAM starts above 32-bit, expand the zone to the maximum
@@ -450,6 +489,9 @@ void __init bootmem_init(void)
 	if (IS_ENABLED(CONFIG_ZONE_DMA) || IS_ENABLED(CONFIG_ZONE_DMA32))
 		reserve_crashkernel();
 
+#ifdef CONFIG_QUICK_KEXEC
+	reserve_quick_kexec();
+#endif
 	memblock_dump_all();
 }
 
